src/Stratis.Bitcoin.Features.BlockStore/AddressIndexing/AddressIndexCache.cs

```diff
@@ -0,0 +1,192 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using LiteDB;
+
+namespace Stratis.Bitcoin.Features.BlockStore.AddressIndexing
+{
+    public class AddressIndexCache
+    {
+        /// <summary>Default maximum number of items (addresses and balance changes) in the cache.</summary>
+        public const int AddressIndexCacheMaxItemsDefault = 100000;
+
+        /// <summary>The number of records to be batched up and written out at once during cache flushes.</summary>
+        private const int BatchSize = 100;
+
+        /// <summary>Maximum number of items (addresses and balance changes) in the cache.</summary>
+        public readonly int MaxItems;
+
+        /// <summary>Number of address index records & associated balance changes present in the cache.</summary>
+        private int itemCount;
+
+        /// <summary>All access to the cache must be protected with <see cref="lockObj"/>.</summary>
+        private readonly Dictionary<string, AddressIndexerData> cachedAddresses;
+
+        /// <summary>The dirty address cache contains the set of addresses that have been
+        /// modified since their inclusion into the cache itself. This speeds up the
+        /// flushing process as only dirty addresses need to be flushed.</summary>
+        /// <remarks>All access to the dirty address cache must be protected with <see cref="lockObj"/>.</remarks>
+        private readonly HashSet<string> dirtyAddresses;
+
+        private object lockObj;
+
+        private readonly LiteDatabase db;
+
+        private readonly LiteCollection<AddressIndexerData> addressIndexerData;
+
+        public AddressIndexCache(LiteDatabase db, string addressIndexerCollectionName)
+        {
+            this.lockObj = new object();
+            this.db = db;
+            this.addressIndexerData = this.db.GetCollection<AddressIndexerData>(addressIndexerCollectionName);
```

Comment:
Kevin can you rename this to `addressIndexerDataCollection` as its a bit confusing..